subsys/dfu/boot/mcuboot_shell.c

```diff
@@ -0,0 +1,202 @@
+
+struct area_desc {
+	const char *name;
+	unsigned int id;
+};
+
+static const struct area_desc areas[] = {
+	{"primary", FLASH_AREA_IMAGE_PRIMARY},
+#ifdef FLASH_AREA_IMAGE_SECONDARY
+	{"secondary", FLASH_AREA_IMAGE_SECONDARY},
+#endif
+};
+
+static const char *swap_state_magic_str(uint8_t magic)
+{
+	switch (magic) {
+	case BOOT_MAGIC_GOOD:
+		return "good";
+	case BOOT_MAGIC_BAD:
+		return "bad";
+	case BOOT_MAGIC_UNSET:
+		return "unset";
+	case BOOT_MAGIC_ANY:
+		return "any";
+	case BOOT_MAGIC_NOTGOOD:
+		return "notgood";
+	}
+
+	return "unknown";
+}
+
+static const char *swap_type_str(uint8_t type)
+{
+	switch (type) {
+	case BOOT_SWAP_TYPE_NONE:
+		return "none";
+	case BOOT_SWAP_TYPE_TEST:
+		return "test";
+	case BOOT_SWAP_TYPE_PERM:
+		return "perm";
+	case BOOT_SWAP_TYPE_REVERT:
+		return "revert";
+	case BOOT_SWAP_TYPE_FAIL:
+		return "fail";
+	}
+
+	return "unknown";
+}
+
+static const char *swap_state_flag_str(uint8_t flag)
+{
+	switch (flag) {
+	case BOOT_FLAG_SET:
+		return "set";
+	case BOOT_FLAG_BAD:
+		return "bad";
+	case BOOT_FLAG_UNSET:
+		return "unset";
+	case BOOT_FLAG_ANY:
+		return "any";
+	}
+
+	return "unknown";
+}
+
+static int cmd_mcuboot_erase(const struct shell *shell, size_t argc,
+			     char **argv)
+{
+	unsigned int id;
+	int err;
+
+	id = strtoul(argv[1], NULL, 0);
+
+	err = boot_erase_img_bank(id);
+	if (err) {
+		shell_error(shell, "failed to erase bank %u", id);
+		return err;
+	}
+
+	return 0;
+}
+
+static int cmd_mcuboot_confirm(const struct shell *shell, size_t argc,
+			       char **argv)
+{
+	int err;
+
+	err = boot_write_img_confirmed();
+	if (err) {
+		shell_error(shell, "failed to confirm: %d", err);
+	}
+
+	return err;
+}
+
+static int cmd_mcuboot_request_upgrade(const struct shell *shell, size_t argc,
+				       char **argv)
+{
+	int permanent = 0;
+	int err;
+
+	if (argc > 1) {
+		if (!strcmp(argv[1], "permanent")) {
+			permanent = 1;
+		} else {
+			shell_warn(shell, "invalid argument!");
+			return -EINVAL;
+		}
+	}
+
+	err = boot_request_upgrade(permanent);
+	if (err) {
+		shell_error(shell, "failed to request upgrade: %d", err);
+	}
+
+	return err;
+}
+
+static int cmd_mcuboot_info_area(const struct shell *shell,
+				 const struct area_desc *area)
+{
+	struct mcuboot_img_header hdr;
+	struct boot_swap_state swap_state;
+	int err;
+
+	err = boot_read_bank_header(area->id, &hdr, sizeof(hdr));
+	if (err) {
+		shell_error(shell, "failed to read %s area (%u) %s: %d",
+			    area->name, area->id, "header", err);
+		return err;
+	}
+
+	shell_print(shell, "%s area (%u):", area->name, area->id);
+	shell_print(shell, "  version: %u.%u.%u+%u",
+		    (unsigned int) hdr.h.v1.sem_ver.major,
+		    (unsigned int) hdr.h.v1.sem_ver.minor,
+		    (unsigned int) hdr.h.v1.sem_ver.revision,
+		    (unsigned int) hdr.h.v1.sem_ver.build_num);
+	shell_print(shell, "  image size: %u",
+		    (unsigned int) hdr.h.v1.image_size);
+
+	err = boot_read_swap_state_by_id(area->id, &swap_state);
```

Comment:
boot_read_swap_state_by_id() was never before considered as API function. It was non static just because migt be build form mcuboot or zephyr surcecode. Value of `struct boot_swap_state` is used to determine the swap-status by int `boot_swap_type_multi(int image_index)` and boot_swap_type().
In #30370 i swich dfu/boot library to use mcuboot surce code - so `boot_swap_type_multi(int image_index)` doesn't present in the API. Of course it might, but personally I belive that boot_swap_type fucntions are more suitable here as user is rather interested in current swap type, and not how to determine it form boot_swap_status structure fields values.

Reply:
I mean to use int mcuboot_swap_type(void) and I can introduced `int mcuboot_swap_type_multi(int image_index)`.

Reply:
@mniestroj what do you think about my concerns?

Reply:
I've used `struct boot_swap_state` for debugging firmware upgrade process when having some problems, such as https://github.com/mcu-tools/mcuboot/issues/829. I agree that in "normal cases" user should not need such information, which is even kind of confusing when user is unfamiliar with mcubot internals.

Keep in mind that `mcuboot_shell.c` already prints information from `mcuboot_swap_type()`. So the question now is whether we can remove this per-slot information about image trailers:
```
  magic: unset
  swap type: none
  copy done: unset
  image ok: unset
```

I would leave that information if easily available (which is exactly why I have that exposed in this PR, as we had non-static functions already). If however functions currently used to fetch this information are not intended to be public (from mcuboot perspective), then let's remove above information. I don't have strong opinion here.

Reply:
I decided to try current command  behavior https://github.com/zephyrproject-rtos/zephyr/pull/30370, see `int mcuboot_read_swap_state_by_id()`.